proxy/apifilterproxy/manager/filter_manager.go

```diff
@@ -0,0 +1,199 @@
+package manager
+
+import (
+	"encoding/json"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/rancher/websocket-proxy/proxy/apifilterproxy/filters"
+	//to register http filter
+	_ "github.com/rancher/websocket-proxy/proxy/apifilterproxy/filters/http"
+	//to register auth filter
+	_ "github.com/rancher/websocket-proxy/proxy/apifilterproxy/filters/auth"
+	"github.com/rancher/websocket-proxy/proxy/apifilterproxy/model"
+	"github.com/rancher/websocket-proxy/proxy/apifilterproxy/util"
+)
+
+var (
+	configFile         string
+	CattleURL          string
+	DefaultDestination string
+	ConfigFields       ConfigFileFields
+	//PathPreFilters is the map storing path -> prefilters[]
+	PathPreFilters map[string][]model.FilterData
+	//PathDestinations is the map storing path -> prefilters[]
+	PathDestinations    map[string]Destination
+	refreshReqChannel   *chan int
+	APIFilterProxyReady bool
+)
+
+//Destination defines the properties of a Destination
+type Destination struct {
+	DestinationURL string   `json:"destinationURL"`
+	Paths          []string `json:"paths"`
+}
+
+//ConfigFileFields stores filter config
+type ConfigFileFields struct {
+	Prefilters   []model.FilterData
+	Destinations []Destination
+}
+
+//SetConfig sets the parameters necessary and initializes API Proxy handler
+func SetConfig(filterConfigFile string, cattleAddr string) {
+	configFile = filterConfigFile
+
+	if configFile == "" {
+		log.Warnf("No path found to the APIfilter config.json file")
+		APIFilterProxyReady = false
+		return
+	}
+
+	CattleURL = "http://" + cattleAddr
+	if len(CattleURL) == 0 {
+		log.Warnf("No CattleAddr set to forward the requests to Cattle")
+		APIFilterProxyReady = false
+		return
+	}
+
+	DefaultDestination = CattleURL
+	refChan := make(chan int, 1)
+	refreshReqChannel = &refChan
+
+	ConfigFields = ConfigFileFields{}
+	PathPreFilters = make(map[string][]model.FilterData)
+	PathDestinations = make(map[string]Destination)
+	err := Reload()
+	if err != nil {
+		log.Warnf("Disabling API filter proxy, failed to load the API filter proxy config with error: %v", err)
+		APIFilterProxyReady = false
+		return
+	}
+
+	APIFilterProxyReady = true
+	log.Infof("Configured the API filter proxy with config %v", filterConfigFile)
+}
+
+func Reload() error {
+	//put msg on channel, so that any other request can wait
+	select {
+	case *refreshReqChannel <- 1:
```

Comment:
Don't use a channel here. Use a RWMutex. All writes to properties should obtain a write lock and then all reads (meaning the http handlers that use the properties) should obtain a read lock.